File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
def check(x):
    ct = 0
    for i in arr:
        if i > x:
            ct += i - x
    return ct >= m

n, m = map(int, input().split())
arr = list(map(int, input().split()))

st = 0
en = 2000000000

while st < en:
    mid = (st+en+1) // 2
    if check(mid):
        st = mid
    else:
        en = mid - 1

print(st)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, m;
    static int[] arr;

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(input.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n];
        st = new StringTokenizer(input.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        int l = 0;
        int r = Integer.MAX_VALUE;

        while (l < r) {
            int mid = 1 + l + (r - l) / 2;
            if (check(mid)) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }

        System.out.println(l);
    }

    static boolean check(int x) {
        long ct = 0;
        for (int a : arr) {
            if (a > x) {
                ct += a - x;
            }
        }

        return  ct >= m;
    }
}
